src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -58,17 +59,24 @@ object RuleUtils {
     val hybridScanDeleteEnabled = HyperspaceConf.hybridScanDeleteEnabled(spark)
 
     def signatureValid(entry: IndexLogEntry): Boolean = {
+      entry.getTagValue(plan, IndexLogEntryTags.SIGNATURE_MATCHED).foreach { sigMatched =>
+        return sigMatched
+      }
+
       val sourcePlanSignatures = entry.source.plan.properties.fingerprint.properties.signatures
       assert(sourcePlanSignatures.length == 1)
       val sourcePlanSignature = sourcePlanSignatures.head
-      signatureMap.getOrElseUpdate(
+
+      val sigMatched = signatureMap.getOrElseUpdate(
         sourcePlanSignature.provider,
         LogicalPlanSignatureProvider
           .create(sourcePlanSignature.provider)
           .signature(plan)) match {
         case Some(s) => s.equals(sourcePlanSignature.value)
         case None => false
       }
+      entry.setTagValue(plan, IndexLogEntryTags.SIGNATURE_MATCHED, sigMatched)
+      sigMatched
```

Comment:
Now that I see this pattern is repeating, does it make sense to introduce something like the following:
```scala
def withCachedTag[A](entry: IndexLogEntry, plan: LogicalPlan, tag: IndexLogEntryTag[A])(
    f: () => A): A = {
  entry.getTagValue(plan, tag).foreach { return _ }
  val ret = f()
  entry.setTagValue(plan, tag, ret)
  ret
}
```
, so you can call it like 
```scala
withCachedTag(index, plan, IndexLogEntryTags.IS_HYBRIDSCAN_CANDIDATE) {
  () => isHybridScanCandidate(index, inputSourceFiles, totalSizeInBytes)
}
```
or even push this helper to `IndexLogEntry`:
```scala
index.withCachedTag(plan, IndexLogEntryTags.IS_HYBRIDSCAN_CANDIDATE) {
  () => isHybridScanCandidate(index, inputSourceFiles, totalSizeInBytes)
}
```